badge-maker/lib/badge-renderers.js

```diff
@@ -270,25 +180,188 @@ class Badge {
     this.label = label
     this.message = message
     this.accessibleText = accessibleText
-    this.renderedLogo = renderedLogo
-    this.renderedLabel = renderedLabel
-    this.renderedMessage = renderedMessage
+    this.logo = logo
+    this.logoWidth = logoWidth
+
+    this.logoElement = this.getLogoElement()
+    this.foregroundGroupElement = this.getForegroundGroupElement()
   }
 
   static render(params) {
     return new this(params).render()
   }
 
+  getLogoElement() {
+    const logoHeight = 14
+    if (!this.hasLogo) {
+      return new NullElement()
+    }
+    return new XmlElement({
+      name: 'image',
+      attrs: {
+        x: this.horizPadding,
+        y: 0.5 * (this.constructor.height - logoHeight),
+        width: this.logoWidth,
+        height: logoHeight,
+        'xlink:href': this.logo,
+      },
+    })
+  }
+
+  getTextElement({ leftMargin, content, link, color, textWidth, linkWidth }) {
+    if (!content.length) {
+      return new NullElement()
+    }
+
+    const { textColor, shadowColor } = colorsForBackground(color)
+    const x =
+      FONT_SCALE_UP_FACTOR * (leftMargin + 0.5 * textWidth + this.horizPadding)
+
+    const text = new XmlElement({
+      name: 'text',
+      content: [content],
+      attrs: {
+        x,
+        y: 140 + this.constructor.verticalMargin,
+        transform: FONT_SCALE_DOWN_VALUE,
+        fill: textColor,
+        textLength: FONT_SCALE_UP_FACTOR * textWidth,
+      },
+    })
+
+    const shadowText = new XmlElement({
+      name: 'text',
+      content: [content],
+      attrs: {
+        'aria-hidden': 'true',
+        x,
+        y: 150 + this.constructor.verticalMargin,
+        fill: shadowColor,
+        'fill-opacity': '.3',
+        transform: FONT_SCALE_DOWN_VALUE,
+        textLength: FONT_SCALE_UP_FACTOR * textWidth,
+      },
+    })
+    const shadow = this.constructor.shadow ? shadowText : new NullElement()
+
+    if (!link) {
+      return new ElementList({ content: [shadow, text] })
+    }
+
+    const rect = new XmlElement({
+      name: 'rect',
+      attrs: {
+        width: linkWidth,
+        x: leftMargin > 1 ? leftMargin + 1 : 0,
+        height: this.constructor.height,
+        fill: 'rgba(0,0,0,0)',
+      },
+    })
+    return new XmlElement({
+      name: 'a',
+      content: [rect, shadow, text],
+      attrs: { target: '_blank', 'xlink:href': link },
+    })
+  }
+
+  getLabelElement() {
+    const leftLink = this.links[0]
+    return this.getTextElement({
+      leftMargin: this.labelMargin,
+      content: this.label,
+      link: !shouldWrapBodyWithLink({ links: this.links }) && leftLink,
```

Comment:
I accepted this because it looked kinda sensible. Then fortunately a test failed :)
`link` here isn't a boolean - its the link URL. Order matters
so this is behaving more like `link: !shouldWrapBodyWithLink({ links: this.links }) ? leftLink : undefined`
Since we both looked at that and thought it was a bool, maybe writing it like that would be clearer?

Reply:
It looks like this made the tests fail? It's not at all clear to me why that would be the case

Reply:
oops, comments crossed 

> Since we both looked at that and thought it was a bool, maybe writing it like that would be clearer?

Yeah that makes sense, it's currently subtle/easy to miss clearly :smile:

Reply:
will go ahead and re-:+1: since this could certainly be deferred like some of other changes mentioned but happy to circle back and do so again if you want to tack that onto this pr